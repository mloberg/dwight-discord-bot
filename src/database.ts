import Keyv from 'keyv';
import { KeyvFile } from 'keyv-file';

import config from './config';
import logger from './logger';

export class Database {
    private impl?: Keyv;

    constructor(private readonly uri: string | undefined, private log: typeof logger) {}

    async get<T>(key: string, fallback?: T): Promise<T> {
        return (await this.kv.get(key)) || fallback;
    }

    async set<T>(key: string, value: T, ttl?: number): Promise<void> {
        await this.kv.set(key, value, ttl);
    }

    async delete(key: string): Promise<void> {
        await this.kv.delete(key);
    }

    async clear(): Promise<void> {
        await this.kv.clear();
    }

    private get kv() {
        if (this.impl) {
            return this.impl;
        }

        this.impl = this.uri?.startsWith('file:')
            ? new Keyv({
                  store: new KeyvFile({
                      filename: this.uri.slice(5),
                  }),
              })
            : new Keyv(config.dbUrl);

        this.impl.on('error', this.log.error);

        return this.impl;
    }
}

export default new Database(config.dbUrl, logger);
