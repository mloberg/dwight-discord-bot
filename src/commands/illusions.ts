import { SlashCommandBuilder } from '@discordjs/builders';
import { CommandInteraction } from 'discord.js';
import { sampleSize } from 'lodash';

import database from '../database';
import { FriendlyError } from '../error';
import { roll } from '../utils';

const deck = [
    'Red dragon',
    'Knight and four guards',
    'Succubus or incubus',
    'Druid',
    'Cloud giant',
    'Ettin',
    'Bugbear',
    'Goblin',
    'Beholder',
    'Archmage and mage apprentice',
    'Night hag',
    'Assassin',
    'Fire giant',
    'Ogre mage',
    'Gnoll',
    'Kobold',
    'Lich',
    'Priest and two acolytes',
    'Medusa',
    'Veteran',
    'Frost giant',
    'Troll',
    'Hobgoblin',
    'Goblin',
    'Iron golem',
    'Bandit captain and three bandits',
    'Erinyes',
    'Berserker',
    'Hill giant',
    'Ogre',
    'Orc',
    'Kobold',
    'You',
    'You',
];

export default {
    config: new SlashCommandBuilder()
        .setName('illusions')
        .setDescription('Manage a Deck of Illusions')
        .addSubcommand((option) =>
            option
                .setName('new')
                .setDescription('Create a new Deck of Illusions')
                .addIntegerOption((option) =>
                    option.setName('missing').setDescription('Number of missing cards (usually d20-1)'),
                ),
        )
        .addSubcommand((option) => option.setName('pull').setDescription('Pull a card from the deck')),
    async handle(command: CommandInteraction): Promise<void> {
        const sub = command.options.getSubcommand();
        const key = `illusions:${command.guild?.id}-${command.user.id}`;

        if (sub === 'new') {
            const missing = command.options.getInteger('missing') ?? roll('d20-1');

            await database.set(key, sampleSize(deck, deck.length - missing));
            return;
        }

        const cards: string[] = await database.get(key, []);
        const card = cards.pop();
        if (!card) {
            throw new FriendlyError('Unable to draw a card.');
        }

        await database.set(key, cards);
        await command.reply(card);
    },
};
