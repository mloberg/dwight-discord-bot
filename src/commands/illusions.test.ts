import { CommandInteraction } from 'discord.js';

import database from '../database';
import { FriendlyError } from '../error';
import illusions from './illusions';

jest.mock('discord.js', () => ({
    CommandInteraction: jest.fn().mockImplementation(() => ({
        guild: {
            id: '1234',
        },
        user: {
            id: '6789',
        },
        reply: jest.fn(),
        options: {
            getSubcommand: jest.fn(),
            getInteger: jest.fn(),
        },
    })),
}));
jest.mock('../database');

const mockDatabase = jest.mocked(database);

describe('/illusions', () => {
    beforeEach(() => {
        mockDatabase.get.mockReset();
        mockDatabase.set.mockReset();
        mockDatabase.delete.mockReset();
    });

    it('is a slash command', () => {
        expect(illusions).toMatchSnapshot();
    });

    describe('new', () => {
        it('creates a deck of illusions', async () => {
            const command = jest.mocked(new CommandInteraction({} as never, {} as never), true);
            command.options.getSubcommand.mockReturnValue('new');
            command.options.getInteger.mockReturnValue(0);

            await illusions.handle(command);
            expect(mockDatabase.set).toHaveBeenCalledWith(
                'illusions:1234-6789',
                expect.arrayContaining(['You', 'Goblin']),
            );
        });

        it('creates a deck of illusions with missing cards', async () => {
            const command = jest.mocked(new CommandInteraction({} as never, {} as never), true);
            command.options.getSubcommand.mockReturnValue('new');

            await illusions.handle(command);
            expect(mockDatabase.set).toHaveBeenCalledWith('illusions:1234-6789', expect.any(Array));
            expect((mockDatabase.set.mock.calls[0][1] as string[]).length).toBeLessThanOrEqual(34);
        });
    });

    describe('pull', () => {
        it('pulls a card', async () => {
            mockDatabase.get.mockResolvedValue(['Goblin', 'Orc', 'You']);

            const command = jest.mocked(new CommandInteraction({} as never, {} as never));

            await illusions.handle(command);
            expect(command.reply).toHaveBeenCalledWith('You');
            expect(mockDatabase.set).toHaveBeenCalledWith('illusions:1234-6789', ['Goblin', 'Orc']);
        });

        it('throws error if no cards can be drawn', async () => {
            mockDatabase.get.mockResolvedValue([]);

            const command = jest.mocked(new CommandInteraction({} as never, {} as never), true);

            await expect(illusions.handle(command)).rejects.toMatchError(new FriendlyError('Unable to draw a card.'));
        });
    });
});
