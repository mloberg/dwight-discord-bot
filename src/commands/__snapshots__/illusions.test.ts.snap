// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/illusions is a slash command 1`] = `
Object {
  "config": Object {
    "default_permission": undefined,
    "description": "Manage a Deck of Illusions",
    "name": "illusions",
    "options": Array [
      Object {
        "description": "Create a new Deck of Illusions",
        "name": "new",
        "options": Array [
          Object {
            "choices": undefined,
            "description": "Number of missing cards (usually d20-1)",
            "name": "missing",
            "required": false,
            "type": 4,
          },
        ],
        "type": 1,
      },
      Object {
        "description": "Pull a card from the deck",
        "name": "pull",
        "options": Array [],
        "type": 1,
      },
    ],
  },
  "handle": [Function],
}
`;
